Export a lora model

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #12724
* __->__ #11045
* #11044

^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

lucylq committed:
Export a lora model

^

Program+data combined currently, using the lora linear definition.

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

digantdesai commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -235,6 +235,18 @@ def build_args_parser() -> argparse.ArgumentParser:
     )
 
+    parser.add_argument(
+        "--adapter_checkpoint",
```
Nit
```suggestion
        "--lora_adapter_checkpoint",
```

digantdesai commented on examples/models/llama/model.py:
```diff
@@ -130,6 +133,21 @@ def __init__(self, **kwargs):
+        adapter_checkpoint = {}
+        adapter_config = {}
+        adapter_checkpoint_path = kwargs.get("adapter_checkpoint", None)
+        if adapter_checkpoint_path:
```
nit: use os.exists or similar from Path?

digantdesai reviewed (commented):
tests?

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

JacobSzwejbka commented on extension/llm/export/config/llm_config.py:
```diff
@@ -90,6 +96,8 @@ class BaseConfig:
     params: Optional[str] = None
     checkpoint: Optional[str] = None
     checkpoint_dir: Optional[str] = None
+    adapter_checkpoint: Optional[str] = None
```
can you name these the same as their corresponding foundational weight fields? I guess thats params and checkpoint? So call it lora_params and lora_checkpoint?

lucylq commented on extension/llm/export/config/llm_config.py:
```diff
@@ -73,10 +73,16 @@ class BaseConfig:
         metadata: Json string containing metadata information.
             e.g. '"{\"get_bos_id\":128000, \"get_eos_ids\":[128009, 128001]}"'
-        use_lora: Rank of the LoRA, if set to 0 then this means no LoRA. For use with QAT.
+        use_lora: Only for use with QAT. Rank of the LoRA adapter, disabled
```
cc @cccclai ?

cccclai replied:
Didn't add it myself either, I think it's likely from Lunwen - I believe it's for the llama 3.2 1b QAT checkpoint which include LoRA, so make sure we don't break llama3.2 QAT model if we use this flag somewhere else

cccclai replied:
https://github.com/pytorch/executorch/pull/5823

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]

lucylq committed:
Update on "Export a lora model"


^

Program+data combined currently, using the lora linear definition.
Issue: https://github.com/pytorch/executorch/issues/9580

Differential Revision: [D75153377](https://our.internmc.facebook.com/intern/diff/D75153377/)

[ghstack-poisoned]
